core/src/main/java/com/linecorp/armeria/common/multipart/MultipartDecoder.java

```diff
@@ -16,43 +16,69 @@
 
 package com.linecorp.armeria.common.multipart;
 
+import static java.util.Objects.requireNonNull;
+
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
 
 import org.reactivestreams.Subscriber;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
+import org.reactivestreams.Subscription;
+
+import com.google.common.math.LongMath;
 
 import com.linecorp.armeria.common.HttpData;
 import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.common.stream.AbortedStreamException;
+import com.linecorp.armeria.common.stream.CancelledSubscriptionException;
+import com.linecorp.armeria.common.stream.DefaultStreamMessage;
 import com.linecorp.armeria.common.stream.HttpDecoder;
 import com.linecorp.armeria.common.stream.HttpDecoderInput;
 import com.linecorp.armeria.common.stream.HttpDecoderOutput;
 import com.linecorp.armeria.common.stream.StreamMessage;
 import com.linecorp.armeria.common.stream.SubscriptionOption;
 import com.linecorp.armeria.internal.common.stream.DecodedHttpStreamMessage;
+import com.linecorp.armeria.internal.common.stream.NoopSubscription;
 
 import io.netty.buffer.ByteBufAllocator;
 import io.netty.util.concurrent.EventExecutor;
 
 final class MultipartDecoder implements StreamMessage<BodyPart>, HttpDecoder<BodyPart> {
 
-    private static final Logger logger = LoggerFactory.getLogger(MultipartDecoder.class);
+    private static final AtomicIntegerFieldUpdater<MultipartDecoder>
+            subscribedUpdater = AtomicIntegerFieldUpdater.newUpdater(MultipartDecoder.class, "subscribed");
 
-    private final StreamMessage<BodyPart> decoded;
+    private final DecodedHttpStreamMessage<BodyPart> decoded;
     private final String boundary;
+    private final CompletableFuture<Void> subscribedCompleteFuture = new CompletableFuture<>();
```

Comment:
I guess we don't need this anymore.

Reply:
I see, so DecodedHttpStreamMessage's abort & subscribe are safe to be called in different thread concurrently, right?

Reply:
I think so because the `subscriptionUpdater` in the `decoded` handles it. 😄

Reply:
Ah I mean if i remove this(also the previous change of after complete), the abort & subscribe may trigger by different thread, so it means DecodedHttpStreamMessage need to handle by self, curious if it's ok to do it. Looks like it's handled correctly, just double confirm.